crashtracker/src/collector_windows/api.rs

```diff
@@ -0,0 +1,789 @@
+    let len = unsafe { read_unaligned(addr_of!((*raw_ptr).len)) };
+
+    anyhow::ensure!(ptr != 0 && len > 0, "Invalid WER context");
+
+    // Read the memory in the target process pointed by ptr
+    let buffer = read_memory_raw(process_handle, ptr as u64, len)?;
+
+    // Copy the buffer into a new Box<[u8]>
+    let boxed_slice = buffer.into_boxed_slice();
+
+    // Leak the Box<[u8]> to get a static reference
+    let static_slice = Box::leak(boxed_slice);
+
+    // Create a new WerContext with the leaked static reference
+    let wer_context = WerContext {
+        prefix: WER_CONTEXT_PREFIX,
+        ptr: static_slice.as_ptr() as usize,
+        len,
+        suffix: WER_CONTEXT_SUFFIX,
+    };
+
+    Ok(wer_context)
+}
+
+// https://github.com/microsoft/win32metadata/issues/1044
+#[repr(align(16))]
+#[derive(Default)]
+struct AlignedContext {
+    ctx: CONTEXT,
+}
+
+fn walk_thread_stack(
+    process_handle: HANDLE,
+    thread_id: u32,
+    modules: &[ModuleInfo],
+) -> Result<StackTrace> {
+    let mut stacktrace = StackTrace::new_incomplete();
+    let thread_handle = unsafe { OpenThread(THREAD_ALL_ACCESS, false, thread_id)? };
+    let mut context = AlignedContext::default();
+
+    #[cfg(target_arch = "x86_64")]
+    {
+        context.ctx.ContextFlags = CONTEXT_FULL_AMD64;
+    }
+    #[cfg(target_arch = "x86")]
+    {
+        context.ctx.ContextFlags = CONTEXT_FULL_X86;
+    }
+
+    unsafe { GetThreadContext(thread_handle, &mut context.ctx)? };
+
+    let mut native_frame = STACKFRAME_EX::default();
+    let machine_type: u32;
+
+    #[cfg(target_arch = "x86_64")]
+    {
+        machine_type = IMAGE_FILE_MACHINE_AMD64.0 as u32;
+        native_frame.AddrPC.Offset = context.ctx.Rip;
+        native_frame.AddrPC.Mode = AddrModeFlat;
+        native_frame.AddrStack.Offset = context.ctx.Rsp;
+        native_frame.AddrStack.Mode = AddrModeFlat;
+        native_frame.AddrFrame.Offset = context.ctx.Rbp;
+        native_frame.AddrFrame.Mode = AddrModeFlat;
+    }
+
+    #[cfg(target_arch = "x86")]
+    {
+        machine_type = IMAGE_FILE_MACHINE_I386.0 as u32;
+        native_frame.AddrPC.Offset = context.ctx.Eip as u64;
+        native_frame.AddrPC.Mode = AddrModeFlat;
+        native_frame.AddrStack.Offset = context.ctx.Esp as u64;
+        native_frame.AddrStack.Mode = AddrModeFlat;
+        native_frame.AddrFrame.Offset = context.ctx.Ebp as u64;
+        native_frame.AddrFrame.Mode = AddrModeFlat;
+    }
+
+    while let TRUE = unsafe {
+        StackWalkEx(
+            machine_type,
+            process_handle,
+            thread_handle,
+            &mut native_frame,
+            &mut context as *mut _ as *mut c_void,
+            None,
+            None,
+            None,
+            None,
+            SYM_STKWALK_DEFAULT,
+        )
+    } {
+        let mut frame = StackFrame::new();
+
+        frame.ip = Some(format!("{:x}", native_frame.AddrPC.Offset));
```

Comment:
Do you want to specify a width here?

Reply:
I don't understand, why would I? I'm just converting the ip to hex

Reply:
I think he means to have consistent padding, like always 16 hex chars.

Reply:
Oh 🤔 Those values are not meant to be interpreted by a human, so I don't think the padding makes sense